File: 20221114-GreastestCommonNumber-java/app/src/main/java/GreatestCommonNumber.java
// [2,6,8,14]	168
//[1,2,3]	6
// n개의 배열에 들은 공통의 최소공배수
// 최소공배수 ?
// 10 5 => 10
// 8 6 = 24
public class GreatestCommonNumber {
  public int solution(int[] arr) {
    int startNumber = 1;
    int temporaryGreatestCommon = 0;

    for(int a : arr) {
      int big = Math.max(a,startNumber);
      int small = Math.min(a,startNumber);

        temporaryGreatestCommon = getGreatestCommonNumber(big, small);

        startNumber = temporaryGreatestCommon;
      }

    int answer = temporaryGreatestCommon;
    return answer;
  }

  public int getGreatestCommonNumber(int big , int small) {
    int a = big;
    int b = small;

    int r = -1;

    while(r != 0) {
      r = big % small;
      big = small;
      small = r;
    }
    return (a * b) / big;
  }
}

File: 20221115-GreatestCommonNumber-javaScript/GreatestCommonNumber.js
// 이해
// 배열에 있는 수들의 최소 공배수 구하기
// 풀이
// 최소 공배수란 두수의 곱을 최대공약수로 나눈 것 과 같음
// 최대공약수 ? -> 유클리드 호제법
// 두 수를 서로 나누고 다시 나눈 나머지와 두 수 중 작은수를 계속해서 나누어 나머지가 0이 될 때의 작은수

export default function solution(arr) {

  function greatestCommonDivisor(a, b) {
    return a % b ? greatestCommonDivisor(b, a % b) : b
  }
  
  return arr.reduce((a, b) => (a*b / greatestCommonDivisor(a,b)))

}